crates/warg/src/lib.rs

```diff
@@ -1,4 +1,154 @@
+    }
+
+    pub fn pbjson_to_prost_timestamp(timestamp: pbjson_types::Timestamp) -> prost_types::Timestamp {
+        prost_types::Timestamp {
+            seconds: timestamp.seconds,
+            nanos: timestamp.nanos,
+        }
+    }
+}
+
+use std::error::Error;
+
+use prost::Message;
+use thiserror::Error;
+
+use hash::HashParseError;
+use signature::Error as SignatureError;
+use signing::SignatureParseError;
+
+/// The envelope struct is used to keep around the original
+/// bytes that the content was serialized into in case
+/// the serialization is not canonical.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct Envelope<Contents> {
+    /// The content represented by content_bytes
+    pub contents: Contents,
+    /// The serialized representation of the content
+    pub content_bytes: Vec<u8>,
+    /// The hash of the key that signed this envelope
+    pub key_id: signing::KeyID,
+    /// The signature for the content_bytes
+    pub signature: signing::Signature,
+}
+
+impl<Contents> Envelope<Contents> {
+    /// Create an envelope for some contents using a signature.
+    pub fn signed_contents(
+        private_key: &signing::PrivateKey,
+        contents: Contents,
+    ) -> Result<Self, SignatureError>
+    where
+        Contents: Signable,
+    {
+        let content_bytes: Vec<u8> = contents.encode();
+
+        let key_id = private_key.public_key().fingerprint();
+        let signature = contents.signature(private_key)?;
+        Ok(Envelope {
+            contents,
+            content_bytes,
+            key_id,
+            signature,
+        })
+    }
+
+    /// Get the representation of the entire envelope as a byte vector.
+    /// This is the logical inverse of `Envelope::from_bytes`.
+    pub fn to_bytes(&self) -> Vec<u8> {
+        let proto_envelope = protobuf::Envelope {
+            contents: self.content_bytes.clone(),
+            key_id: self.key_id.to_string(),
+            signature: self.signature.to_string(),
+        };
+        proto_envelope.encode_to_vec()
+    }
+
+    /// Create an entire envelope from a byte vector.
+    /// This is the logical inverse of `Envelope::as_bytes`.
+    pub fn from_bytes(bytes: Vec<u8>) -> Result<Self, ParseEnvelopeError>
+    where
+        Contents: for<'a> TryFrom<&'a [u8], Error = ErrorBox>,
+    {
+        // Parse outer envelope
+        let envelope = protobuf::Envelope::decode(&*bytes)?;
+        // Parse contents
+        let content_bytes = envelope.contents.clone();
+        let contents = content_bytes
+            .as_slice()
+            .try_into()
+            .map_err(ParseEnvelopeError::ContentsParseError)?;
+        // Read key ID and signature
+        let key_id = envelope.key_id.into();
+        let signature = envelope.signature.parse()?;
+
+        Ok(Envelope {
+            contents,
+            content_bytes,
+            key_id,
+            signature,
+        })
+    }
+}
+
+type ErrorBox = Box<dyn Error + Send + Sync + 'static>;
```

Comment:
Any particular reason to avoid `anyhow` here?

Reply:
Just unfamiliarity, `anyhow::Error` does do the trick nicely though so I've implemented that.